src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs

```diff
@@ -23,11 +23,27 @@ public sealed partial class JsonSerializerOptions
         // The global list of built-in converters that override CanConvert().
         private static JsonConverter[]? s_defaultFactoryConverters;
 
+        // Stores the JsonTypeInfo factory, which requires unreferenced code and must be rooted by the reflection-based serializer.
+        private static Func<Type, JsonSerializerOptions, JsonTypeInfo>? s_typeInfoCreationFunc;
+
+        [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
+        private static void RootReflectionSerializerDependencies()
+        {
+            if (s_defaultSimpleConverters is null)
```

Comment:
I'm not sure if this is under lock or not, but if not would it make sense to check for `s_typeInfoCreationFunc` being null instead? if you check for the first field this might backfire once in a while if first thread starts initializing while other one assumes it's already initialized because first field is set

Reply:
It's ok for contending threads to initialize these fields multiple times, this is a check meant to avoid incurring allocations once initialization has completed.